scipio/src/sys/uring.rs

```diff
@@ -831,7 +795,7 @@ impl Reactor {
     }
 
     pub(crate) fn statx(&self, source: &Source) {
-        let op = match source.source_type() {
+        let op = match &*source.source_type() {
```

Comment:
I trust you to know what you're doing rust wise but I don't understand this. Would like to understand a bit better before pressing the button.

`source_type()` returns a `Ref` from the immutable borrow.
Aren't you getting a naked reference to the interior of the cell by doing `&*` ? I thought this was not possible?

But even if it is possible, wouldn't that live for just as long as the `Ref`?

Reply:
Taking reference to the interior is possible, Ref implements Deref. Moreover, it is essential, otherwise you won‘t be able to call any methods at all on values behind RefCells.

The reference would indeed leave only as long as Ref, but this is enough in this case: rust has temporary lifetime extension rules, much like C++, but you don‘t generally need to understand them in details -- the worst case is that the code fails to compile, and you need to introduce a temporary.

Note that this is exactly the same API as for mutex guards. There, one also often uses &* incantation to summon a short-lived reference to a data protected by a guard.

Reply:
Right, so what is not possible is to generate a reference that escapes the lifetime of the borrow.

Correct?
My first instinct looking at this code is that the lifetime of the borrow would be not the entire match statement but just the first part `match <this_part>`

But I agree that as far as lifetimes go, if it compiled it is correct.

Reply:
Yes, that is correct.

I think all temporaries are live until the end of the full expression (the whole `match` In this case), unless the are extended further by something like `let x = &“hello“.to_string();`. But this is a part of Rust I don’t know and just rely on the compiler to tell me when I am wrong.